core/src/inputs/advice/merkle_path_set.rs

```diff
@@ -0,0 +1,339 @@
+    ///
+    /// Merkle tree of depth 1 has two leaves, depth 2 has four leaves etc.
+    pub fn depth(&self) -> u32 {
+        self.total_depth
+    }
+
+    /// Returns a node at the specified index.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// * The specified index not valid for the depth of structure.
+    /// * Requested node does not exist in the set.
+    pub fn get_node(&self, depth: u32, index: u64) -> Result<Word, AdviceSetError> {
+        if index >= 2u64.pow(self.total_depth) {
+            return Err(AdviceSetError::InvalidIndex(self.total_depth, index));
+        }
+        if depth != self.total_depth {
+            return Err(AdviceSetError::InvalidDepth(self.total_depth, depth));
+        }
+
+        let pos = 2u64.pow(depth) + index;
+        let index = (pos / 2) as u64;
+
+        match self.paths.get(&index) {
+            None => Err(AdviceSetError::NodeNotExist(index)),
+            Some(path) => {
+                if is_even(pos) {
+                    Ok(path[0])
+                } else {
+                    Ok(path[1])
+                }
+            }
+        }
+    }
+
+    /// Returns a Merkle path to the node at the specified index. The node itself is
+    /// not included in the path.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// * The specified index not valid for the depth of structure.
+    /// * Node of the requested path does not exist in the set.
+    pub fn get_path(&self, depth: u32, index: u64) -> Result<Vec<Word>, AdviceSetError> {
+        if index >= 2u64.pow(self.total_depth) {
+            return Err(AdviceSetError::InvalidIndex(self.total_depth, index));
+        }
+        if depth != self.total_depth {
+            return Err(AdviceSetError::InvalidDepth(self.total_depth, depth));
+        }
+
+        let pos = 2u64.pow(depth) + index;
+        let index = pos / 2;
+
+        match self.paths.get(&index) {
+            None => Err(AdviceSetError::NodeNotExist(index)),
+            Some(path) => {
+                let mut local_path = path.clone();
+                if is_even(pos) {
+                    local_path.remove(0);
+                    Ok(local_path)
+                } else {
+                    local_path.remove(1);
+                    Ok(local_path)
+                }
+            }
+        }
+    }
+
+    /// Replaces the leaf at the specified index with the provided value.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// * Requested node does not exist in the set.
+    pub fn update_leaf(&mut self, index: u64, value: Word) -> Result<(), AdviceSetError> {
+        let depth = self.depth();
+        if index >= 2u64.pow(depth) {
+            return Err(AdviceSetError::InvalidIndex(depth, index));
+        }
+        let pos = 2u64.pow(depth) + index;
+        if self.paths.get(&(pos / 2)).is_none() {
+            return Err(AdviceSetError::NodeNotExist(index));
+        }
```

Comment:
You could probably get the path and perform this check in the same block similar to how you did it in other places. For example, something like:
```Rust
let path = match self.paths.get(&(pos / 2)) {
    None => return Err(AdviceSetError::NodeNotExist(index)),
    Some(path) => path.clone(),
};
```